test/roles.js

```diff
@@ -0,0 +1,942 @@
+      let virtualEventLog = JSON.parse(res.virtualTransactions[0].logs);
+      let e = virtualEventLog.events.find(x => x.event === 'rolePayment');
+      assert.ok(e, 'Expected to find rolePayment event');
+
+      refBlockNumber = fixture.getNextRefBlockNumber();
+      transactions = [];
+      transactions.push(new Transaction(refBlockNumber, fixture.getNextTxId(), 'voter2', 'roles', 'disapproveCandidate', '{ "id": "2" }'));
+
+      block = {
+        refHiveBlockNumber: refBlockNumber,
+        refHiveBlockId: 'ABCD1',
+        prevRefHiveBlockId: 'ABCD2',
+        timestamp: '2021-03-13T00:00:03',
+        transactions,
+      };
+      await fixture.sendBlock(block);
+
+      res = (await fixture.database.getLatestBlockInfo());
+      // console.log(res);
+      await tableAsserts.assertNoErrorInLastBlock();
+
+      await assertWeightConsistency(1, 'PRO');
+      await assertWeightConsistency(2, 'PRO');
+      await assertWeightConsistency(3, 'PRO');
+      await assertWeightConsistency(4, 'PRO');
+
+      assert.ok(res.virtualTransactions.length > 0, 'Expected to find virtualTransactions');
+      virtualEventLog = JSON.parse(res.virtualTransactions[0].logs);
+      e = virtualEventLog.events.find(x => x.event === 'rolePayment');
+      assert.ok(e, 'Expected to find rolePayment event');
+
+      // balance asserts
+      await tableAsserts.assertUserBalances({ account: 'organizer', symbol: 'BEE', balance: '25.00000000'});
+      await tableAsserts.assertUserBalances({ account: 'voter1', symbol: 'BEE', balance: '25.00000000'});
+      await tableAsserts.assertUserBalances({ account: 'voter2', symbol: 'BEE', balance: '25.00000000'});
+      await tableAsserts.assertUserBalances({ account: 'voter3', symbol: 'BEE'});
+      await tableAsserts.assertUserBalances({ account: 'voter4', symbol: 'BEE'});
+      await assertContractBalance('roles', 'BEE', '125.00000000');
+
+      // let roleState = await fixture.database.find({
+      //   contract: 'roles',
+      //   table: 'roles',
+      //   query: {}
+      // });
+      // console.log(JSON.stringify(roleState, null, 2));
+
+      refBlockNumber = fixture.getNextRefBlockNumber();
+      transactions = [];
+      transactions.push(new Transaction(refBlockNumber, fixture.getNextTxId(), 'voter2', 'roles', 'approveCandidate', '{ "id": "4" }'));
+
+      block = {
+        refHiveBlockNumber: refBlockNumber,
+        refHiveBlockId: 'ABCD1',
+        prevRefHiveBlockId: 'ABCD2',
+        timestamp: '2021-03-13T00:00:06',
+        transactions,
+      };
+      await fixture.sendBlock(block);
+
+      res = (await fixture.database.getLatestBlockInfo());
+      // console.log(res);
+      await tableAsserts.assertNoErrorInLastBlock();
+      assert.ok(res.virtualTransactions.length == 0, 'Expected to not find virtualTransactions'); 
+
+      resolve();
+    })
+      .then(() => {
+        fixture.tearDown();
+        done();
+      });
+  });  
+
```

Comment:
does this test deposits via dtf/distribution contracts?

Reply:
Not currently, I was going to confirm in testnet